astropy/coordinates/baseframe.py

```diff
@@ -423,40 +423,110 @@ def __init__(self, *args, copy=True, representation_type=None,
         _normalize_representation_type(kwargs)
         representation_type = kwargs.pop('representation_type', representation_type)
 
-        if representation_type is not None or differential_type is not None:
+        self._representation = self._infer_representation_differential_type(representation_type, differential_type)
```

Comment:
Ah, seeing it together it is even nicer than I thought! Really only a nitpick left - my tendency would be to shorten the method names to `_infer_representation(...)` and `_infer_data(...)` (or latter perhaps `_infer_representation_data` - the differential being part of the representation data).

Reply:
Done!